libraries/persistent_store/src/store.rs

```diff
@@ -128,3 +168,1089 @@ impl StoreUpdate {
+    /// Note that the deleted entries are before any compaction, so they may point outside the
+    /// window. This is more expressive than returning the deleted entries after compaction since
+    /// compaction can be controlled independently.
+    pub fn apply(&mut self, operation: &StoreOperation) -> (Vec<StoreHandle>, StoreResult<()>) {
+        let deleted = |store: &Store<BufferStorage>, delete_key: &dyn Fn(usize) -> bool| {
+            store
+                .iter()
+                .unwrap()
+                .map(|x| x.unwrap())
+                .filter(|x| delete_key(x.key as usize))
+                .collect::<Vec<_>>()
+        };
+        match *operation {
+            StoreOperation::Transaction { ref updates } => {
+                let keys: HashSet<usize> = updates.iter().map(|x| x.key()).collect();
+                let deleted = deleted(self, &|key| keys.contains(&key));
+                (deleted, self.transaction(updates))
+            }
+            StoreOperation::Clear { min_key } => {
+                let deleted = deleted(self, &|key| key >= min_key);
+                (deleted, self.clear(min_key))
+            }
+            StoreOperation::Prepare { length } => (Vec::new(), self.prepare(length)),
+        }
+    }
+
+    /// Initializes an erased storage as if it has been erased `cycle` times.
+    pub fn init_with_cycle(storage: &mut BufferStorage, cycle: usize) {
+        let format = Format::new(storage).unwrap();
+        // Write the init info of the first page.
+        let mut index = format.index_init(0);
+        let init_info = format.build_init(InitInfo {
+            cycle: usize_to_nat(cycle),
+            prefix: 0,
+        });
+        storage.write_slice(index, &init_info).unwrap();
+        // Pad the first word of the page. This makes the store looks used, otherwise we may confuse
+        // it with a partially initialized store.
+        index.byte += 2 * format.word_size() as usize;
+        storage
+            .write_slice(index, &vec![0; format.word_size() as usize])
+            .unwrap();
+        // Inform the storage that the pages have been used.
+        for page in 0..storage.num_pages() {
+            storage.set_page_erases(page, cycle);
+        }
+    }
+}
+
+/// Represents an entry in the store.
+#[derive(Debug)]
+enum ParsedEntry {
+    /// Padding entry.
+    ///
+    /// This can be any of the following:
+    /// - A deleted user entry.
+    /// - A completed internal entry.
+    /// - A wiped partial entry.
+    Padding,
+
+    /// Non-deleted user entry.
+    User(Header),
+
+    /// Internal entry.
+    Internal(InternalEntry),
+
+    /// Partial user entry with non-erased footer.
+    ///
+    /// The fact that the footer is not erased and does not checksum, means that the header is
+    /// valid. In particular, the length is valid. We cannot wipe the entry because wiping the
+    /// footer may validate the checksum. So we mark the entry as deleted, which also wipes it.
```

Comment:
Double-checking this comment: "We cannot wipe the entry" does not mean that data may not be wiped, but talks about meta-bytes?

Reply:
> "We cannot wipe the entry" does not mean that data may not be wiped

Yes

> but talks about meta-bytes?

Not sure what you mean. The explanation is that "we cannot wipe the entry" means "wiping the entry before marking the entry as deleted could introduce corruption, so we first have to mark the entry as deleted, then let the usual deletion mechanism wipe the entry."

Reply:
"meta-bytes" was the marker, so my assumption was correct.